crates/gossip/src/gossip.rs

```diff
@@ -26,132 +35,634 @@ impl K2GossipFactory {
+        let mut initiated_lock = state.lock().await;
+        if initiated_lock.is_some() {
+            tracing::info!("initiate_gossip: already initiated");
+            return Ok(false);
+        }
+
+        let Some(target_url) = self
+            .peer_store
+            .get(target.clone())
+            .await?
+            .and_then(|t| t.url.clone())
+        else {
+            tracing::info!("initiate_gossip: target not found: {:?}", target);
+            return Ok(false);
+        };
+
+        let (our_agents, our_arc_set) = self.local_agent_state().await?;
+
+        let new_since = self
+            .peer_meta_store
+            .new_ops_bookmark(target_url.clone())
+            .await?
+            .unwrap_or(UNIX_TIMESTAMP);
+
+        let round_state =
+            GossipRoundState::new(target_url.clone(), our_agents.clone());
+        let initiate = K2GossipInitiateMessage {
+            session_id: round_state.session_id.clone(),
+            participating_agents: encode_agent_ids(our_agents),
+            arc_set: Some(ArcSetMessage {
+                arc_sectors: our_arc_set.into_raw().collect(),
+            }),
+            new_since: new_since.as_micros(),
+            max_new_bytes: self.config.max_gossip_op_bytes,
+        };
+
+        // Right before we send the initiate message, check whether the target has already
+        // initiated with us. If they have, we can just skip initiating.
+        if self
+            .accepted_round_states
+            .lock()
+            .await
+            .contains_key(&target_url)
+        {
+            tracing::info!("initiate_gossip: already accepted");
+            return Ok(false);
+        }
+
+        tracing::trace!("Initiate gossip with {:?}: {:?}", target, initiate);
+
+        send_gossip_message(
+            &self.response_tx,
+            target_url,
+            GossipMessage::Initiate(initiate),
+        )?;
+        *initiated_lock = Some(round_state);
+
+        Ok(true)
+    }
+
     /// Handle a gossip message.
     ///
     /// Produces a response message if the input is acceptable and a response is required.
     /// Otherwise, returns None.
     fn handle_gossip_message(
         &self,
-        from: AgentId,
-        msg: K2GossipMessage,
-    ) -> K2Result<Option<K2GossipMessage>> {
-        println!("handle_gossip_message from: {:?}, msg: {:?}", from, msg);
+        from_peer: Url,
+        msg: GossipMessage,
+    ) -> K2Result<()> {
+        tracing::debug!(
+            "handle_gossip_message from: {:?}, msg: {:?}",
+            from_peer,
+            msg
+        );
 
-        let Some(msg) = msg.gossip_message else {
-            return Err(K2Error::other("no gossip message"));
-        };
+        let this = self.clone();
+        tokio::task::spawn(async move {
+            if let Err(e) = this.respond_to_msg(from_peer, msg).await {
+                tracing::error!("could not respond to gossip message: {:?}", e);
+            }
+        });
```

Comment:
Like an unbounded channel, unlimited tasks here could fill up our memory if the app is running slow. As you say, this module is rough, so doesn't have to be now, but at some point if gossip is overloading the system we need a relief valve to shut it down.

Reply:
We could limit the number of concurrent incoming rounds and just return a busy message? That along with the number of messages that can be sent within a single round would limit the number of tasks that could be open